### ISSUE TITLE
Optimization: construct download link from work id instead of scraping it

### ISSUE THREAD

**Issue by nianeyna** [nianeyna/ao3downloader#186]
Could be up to 2x faster. Problem: filename customization relies on metadata scraped from the work page. In many cases that's fine because the same data is available from the list view, but subscriptions are an issue. Could do a fallback to scraping the work in that case?

**Comment by RowanSkie** [comment]
This can actually help fix potential issues for those using the #220 workaround.

**Comment by nianeyna** [comment]
Yeah, this issue is probably my top priority right now, minimizing hits to ao3 would be a HUGE win.

**Comment by Newgraywolf** [comment]
Looking at this, could instead of checking on the individual fics page, maybe set up a check that asks the user if they want to have it check their subscriptions page on AO3? Then it gives everything it finds from those pages the "subscribed" tag?


It isn't perfect, but it could help mitigate a few requests? I can't see any other way to check what fics someone is subscribed to outside of checking the work, or the subscriptions page, but there may be another way.

**Comment by nianeyna** [comment]
That's the *default* file name pattern 🙂 

Scraping *the file itself* for the metadata is actually an interesting idea that I hadn't considered, though. You'd have to save the file to some sort of temporary location first of course, but... it could be doable. 🤔 I'd have to write a bunch more pdf scraping logic which is always a vile endeavor lol but... worth it perhaps

**Comment by nianeyna** [comment]
Implementation note: constructed download urls must match the real ones *exactly* (including properly transformed/encoded work title and ?updated_at query string). Otherwise, ao3 has to regenerate the download, which can be very slow (and is also rude to them). updated_at timestamp is available in an html comment on listing pages. Work title code is here: https://github.com/otwcode/otwarchive/blob/master/app/models/download.rb#L54

**Comment by nianeyna** [comment]
Investigation notes: Title, author, work id, and updated_at can be scraped from any **work blurb**. Notably this does NOT include subscriptions pages - updated_at appears to be linked to the *blurb*, and subscriptions don't have them. It *does* include blurbs that appear on non-listing pages (home page, dashboards). 

So: if updated_at cannot be found, fall back to scraping the work page instead of trying to construct the download link.

That solves: subscriptions page updated_at, subscriptions page custom file names (since we'll be scraping the work page anyway, we can continue to get title metadata in the same way as the current code). 

Does not solve: other listing custom filenames which include metadata that cannot be found on the listing page.

So: if we are constructing the download link, AND the filename template contains tokens which are not included in listings (this should be based on a hardcoded list, probably - sometimes something won't be available in the listing because it genuinely has no value for that work, and we don't want to waste a fallback in that case) final fallback case to get the required data.

Still need to decide: whether the final fallback should be "scrape the work page" - same flow as the subscription/missing updated_at branch, and probably could be branched off at the same time - or whether it should be "pull data from the actual downloaded file" as Newgraywolf suggested. First one is simpler, second one is probably faster (even file parsing will likely outperform the bot queue wait for an ao3 request).
